osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuMovement.cs

```diff
@@ -0,0 +1,631 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using MathNet.Numerics;
+using MathNet.Numerics.LinearAlgebra;
+using MathNet.Numerics.Interpolation;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Osu.Difficulty.MathUtil;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Preprocessing
+{
+    public class OsuMovement
```

Comment:
This class is mostly obfuscated magic. I don't understand what `coeffs` are trying to do or mean.

Reply:
all the coeffs go into "corrections" which add or subtract from the "index of difficulty" basically just by tweaking them to make results look better (though there is reasoning for each one).

The coeffs stuff is pretty opaque, they are for "correction 0" and "correction 3", which apply to the position of note 0 and note 3. In these numberings, he defines note 2 as the one that we're actually trying to hit, note 0 is two before it and note 3 is one after it.

For each one of these there is a big list of "coeffs" which go into a big formula that defines how the position of 0 and 3 affect the difficulty of hitting note 2 - the position is first transformed to be in units where the prev note (i.e. note 1) is the origin and the current note (note 2) is somewhere on the positive x axis, then it's passed into this big formula.

I tried refactored the coeffs stuff into a different class and give all the parts better names, but it's still confusing, and pretty impossible to visualise in your head. There is external code to visualise all the coeffs stuff somewhere on wolfram cloud, but I haven't used it. 

Ultimately the coeffs stuff is just a way of defining a tweakable function from position of the note to a correction, 

I think it's probably best to replace this entirely with something simpler that gives the same results. Maybe just define some values on a grid of (angle, distance) and interpolate?